site/blocks/KeyInsights.tsx

```diff
@@ -0,0 +1,290 @@
+            )
+        )
+
+        if (selectedSlideIdx === -1) return
+        setSelectedId(selectedSlideIdx.toString())
+    }, [slides])
+
+    // Scroll to selected item
+    useEffect(() => {
+        const item = apiRef.current.getItemById(selectedId)
+        if (!item) return
+
+        apiRef.current.scrollToItem(item, "smooth", "center", "nearest")
+    }, [selectedId])
+
+    // Select active slide when corresponding thumb selected
+    useEffect(() => {
+        if (!slides) return
+
+        // A less imperative, more React way to do this would be preferred. To
+        // switch between slides, I aimed to keep their content untouched
+        // (including event listeners hydrated by other components), while only
+        // updating their wrappers. Managing the switching logic through React
+        // would have required hydrating KeyInsightsSlides as well as all
+        // possible content components within them - even though they would have
+        // been already hydrated at the page level. From that perspective, the
+        // gain is not clear, and the approach not necessarily cleaner, so I
+        // stuck with the imperative approach.
+
+        slides
+            .querySelectorAll(`.${KEY_INSIGHTS_SLIDE_CLASS_NAME}`)
+            .forEach((slide, idx) => {
+                if (idx === Number(selectedId)) {
+                    slide.setAttribute("data-active", "true")
+                    const windowUrl = getWindowUrl()
+                    const anchor = slide.querySelector("h4")?.getAttribute("id")
+                    if (!anchor) return
+                    setWindowUrl(
+                        windowUrl
+                            .updateQueryParams({
+                                [KEY_INSIGHTS_INSIGHT_PARAM]: anchor,
+                            })
+                            // When a key insight slug is changed, links
+                            // pointing to that key insight soft-break and take
+                            // readers to the top of the page. Adding an anchor
+                            // pointing to the the block title (h3) serves as a
+                            // stopgap, taking readers to the key insights block
+                            // instead but without selecting a particular
+                            // insight.
+                            //
+                            // This also improves the UX of readers coming
+                            // through shared insights URL. e.g.
+                            // /key-insights-demo?insight=insight-1#key-insights
+                            // shows the whole insights block, including its
+                            // titles (the target of the anchor).
+                            .update({ hash: `#${slug}` })
+                    )
+                } else {
+                    slide.setAttribute("data-active", "false")
+                }
+            })
+    }, [selectedId])
```

Comment:
I wonder if this would be clearer if we used a little mobx data object and then a function reactive component that renders it an an observer to update the url? I think this implementation is perfectly fine as is, more of a longer term question of how to do this kind of thing

Reply:
Quite possibly, thanks for the pointers.

As for adding `slug` and `slides` to the dependency array, this is possible but requires to set the default for `selectedId` as the first tab in an effect (see [keyinsights-set-default-in-effect.patch.txt](https://github.com/owid/owid-grapher/files/8825032/keyinsights-set-default-in-effect.patch.txt)
 ). This makes the first tab switch from "not selected" to "selected" in a visible way, which is slightly suboptimal given the expected prominence of the "key insights" block on topics page (likely to be visible without scroll).


https://user-images.githubusercontent.com/13406362/171660123-2f34ed95-e7b6-42c6-87c4-dbe509a8a6c3.mov


So for preventing flickering of the first tab (in case none are selected from the URL), I'm leaning towards keeping the default selected tab state set on the server (at the `useState` level ) rather than on the client in an effect. 

But again, happy to consider other approaches if they more direct and simpler to reason with.

Reply:
Makes sense